### PR TITLE
make sure a match between logits and logits_dtype in evaluation_loop

### PR DESCRIPTION
**Pull request by xuwh-intel**
# What does this PR do?

In the trainer.py/evaluation_loop function, `logits = self.preprocess_logits_for_metrics(logits, labels)` will change the type of logits from tuple to tensor.  But the logits_dtype obtained from `get_dtype(logits)` is still a list type and does not match the shape of the new logits. this will cause the following error in `all_preds = convert_into_dtypes(all_preds, logits_dtype)`.  So I re-get dtype of logits after `logits = self.preprocess_logits_for_metrics(logits, labels)` to ensure a match between logits and logits_dtype.

image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
@xuwh-intel Ooh interesting! Could you share a command so that I reproduce this issue please? I would like to understand why it doesn't happen in the CI examples.

**Comment by xuwh-intel** [comment]
> @xuwh-intel Ooh interesting! Could you share a command so that I reproduce this issue please? I would like to understand why it doesn't happen in the CI examples.

It looks like the model I tested does not set `keys_to_ignore_at_inference = ["past_key_values"]` in configuration, which caused this error.

**Comment by regisss** [comment]
@xuwh-intel Is it a model from the Transformers library and that has not been optimized for Gaudi in Optimum Habana?

**Comment by xuwh-intel** [comment]
> @xuwh-intel Is it a model from the Transformers library and that has not been optimized for Gaudi in Optimum Habana?

Yes, it's chatglm2.
